### PR TITLE
The implementation of FindTraceIDs function for ElasticSearch reader.

### PR DESCRIPTION
**Pull request by vlamug**
Related to https://github.com/jaegertracing/jaeger/issues/1241

## Which problem is this PR solving?
It allows to use `FindTraceIDs` function for ElasticSearch storage.

## Short description of the changes
The implementations of `FindTraceIDs` function for ElasticSearch was added.

### CODE REVIEW

File: plugin/storage/es/spanstore/reader.go [hunk 1]
```diff
@@ -234,9 +234,35 @@ func (s *SpanReader) FindTraces(ctx context.Context, traceQuery *spanstore.Trace
 	return s.multiRead(ctx, uniqueTraceIDs, traceQuery.StartTimeMin, traceQuery.StartTimeMax)
 }
 
-// FindTraceIDs is not implemented.
+// FindTraceIDs retrieves traces IDs that match the traceQuery
 func (s *SpanReader) FindTraceIDs(ctx context.Context, traceQuery *spanstore.TraceQueryParameters) ([]model.TraceID, error) {
-	return nil, errors.New("not implemented") // TODO: Implement
+	if err := validateQuery(traceQuery); err != nil {
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: I see a lot of repeated code between `FindTraces` and `FindTraceIDs`, are we able to refactor such that the initial validation and retrieval are shared?

Also, could you create a new span here - similar to L221

File: plugin/storage/es/spanstore/reader.go [hunk 2]
```diff
@@ -234,9 +234,35 @@ func (s *SpanReader) FindTraces(ctx context.Context, traceQuery *spanstore.Trace
...
+	}
+
+	esTraceIDs, err := s.findTraceIDs(ctx, traceQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	traceIDs := []model.TraceID{}
+	for _, ID := range esTraceIDs {
+		if len(traceIDs) >= traceQuery.NumTraces {
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: Why is this required? I see that `findTraceIDs` already takes in `traceQuery.NumTraces`. Does it return a larger number of traces than specified?

File: plugin/storage/es/spanstore/reader.go [hunk 3]
```diff
@@ -234,9 +234,35 @@ func (s *SpanReader) FindTraces(ctx context.Context, traceQuery *spanstore.Trace
...
+			break
+		}
+
+		traceID, err := model.TraceIDFromString(ID)
+		if err != nil {
+			return nil, err
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: nit: Could you log the traceID string?

File: plugin/storage/es/spanstore/reader.go [hunk 4]
```diff
@@ -234,9 +234,35 @@ func (s *SpanReader) FindTraces(ctx context.Context, traceQuery *spanstore.Trace
...
+	if traceQuery.NumTraces == 0 {
+		traceQuery.NumTraces = defaultNumTraces
+	}
+
+	esTraceIDs, err := s.findTraceIDs(ctx, traceQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	traceIDs := []model.TraceID{}
```
Comment (mkabischev) [inline plugin/storage/es/spanstore/reader.go]: You know the length of result slice, so you can preallocate it.

File: plugin/storage/es/spanstore/reader.go (hunk 1 above)
Comment (vlamug) [reply plugin/storage/es/spanstore/reader.go]: Ok, thanks.

File: plugin/storage/es/spanstore/reader.go (hunk 4 above)
Comment (vlamug) [reply plugin/storage/es/spanstore/reader.go]: I agree. Thanks.

File: plugin/storage/es/spanstore/reader.go (hunk 2 above)
Comment (vlamug) [reply plugin/storage/es/spanstore/reader.go]: I studied the implementation of such function for cassandra and there are some troubles with it, therefore the implementation for cassandra uses this condition. It seems that here it is ok. I will remove it.

File: plugin/storage/es/spanstore/reader.go (hunk 3 above)
Comment (vlamug) [reply plugin/storage/es/spanstore/reader.go]: Good suggestion. Thanks.

**Comment by vlamug** [comment]
@vprithvi , @mkabischev , guys, the changes were added. Please, check it. Thanks.

File: plugin/storage/es/spanstore/reader.go [hunk 5]
```diff
@@ -317,6 +329,20 @@ func (s *SpanReader) multiRead(ctx context.Context, traceIDs []string, startTime
 	return traces, nil
 }
 
+func (s *SpanReader) validateQueryAndFindTraceIDs(ctx context.Context, traceQuery *spanstore.TraceQueryParameters) ([]string, error) {
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: Looks good!

File: plugin/storage/es/spanstore/reader_test.go [hunk 6]
```diff
@@ -668,11 +668,142 @@ func TestFindTraceIDs(t *testing.T) {
 	testGet(traceIDAggregation, t)
 }
 
-func TestFindTraceIDNotImplemented(t *testing.T) {
+func TestSpanReader_TestFindTraceIDs(t *testing.T) {
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader_test.go]: These tests look very similar to the `TestSpanReader_TestFindTraces, ...` tests. Is there a way to DRY them up?

File: plugin/storage/es/spanstore/reader.go [hunk 7]
```diff
@@ -317,6 +329,20 @@ func (s *SpanReader) multiRead(ctx context.Context, traceIDs []string, startTime
...
+	childSpan, _ := opentracing.StartSpanFromContext(ctx, "validateQueryAndFindTraceIDs")
```
Comment (pavolloffay) [inline plugin/storage/es/spanstore/reader.go]: I don't think we need a span for this operation.

**Comment by pavolloffay** [comment]
The integration tests can be done in a separate PR. This PR LGTM. It would be nice to reuse tests if possbile though as @vprithvi mentioned

**Comment by vlamug** [comment]
@vprithvi , hi) Could you please check the last update?)
We are waiting for the changes in this pull request in order to use FindTraceIDs function.

A lot of thanks.

File: plugin/storage/es/spanstore/reader.go (hunk 7 above)
Comment (yurishkuro) [reply plugin/storage/es/spanstore/reader.go]: please delete it

File: plugin/storage/es/spanstore/reader.go [hunk 8]
```diff
@@ -377,6 +395,7 @@ func validateQuery(p *spanstore.TraceQueryParameters) error {
 func (s *SpanReader) findTraceIDs(ctx context.Context, traceQuery *spanstore.TraceQueryParameters) ([]string, error) {
 	childSpan, _ := opentracing.StartSpanFromContext(ctx, "findTraceIDs")
 	defer childSpan.Finish()
+
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: nit: extraneous change

File: plugin/storage/es/spanstore/reader.go [hunk 9]
```diff
@@ -333,18 +352,17 @@ func (s *SpanReader) multiRead(ctx context.Context, traceIDs []string, startTime
 				return nil, err
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: I noticed that the integration tests failed. Perhaps L298 requires a `traceID.String()`?

File: plugin/storage/es/spanstore/reader_test.go [hunk 10]
```diff
@@ -702,12 +702,25 @@ func TestFindTraceIDs(t *testing.T) {
 	testGet(traceIDAggregation, t)
 }
 
-func TestFindTraceIDNotImplemented(t *testing.T) {
-	withSpanReader(func(r *spanReaderTest) {
-		traceIDs, err := r.reader.FindTraceIDs(context.Background(), nil)
-		assert.Nil(t, traceIDs)
-		assert.EqualError(t, err, "not implemented")
-	})
+func TestTraceIDsStringsToModelsConvertation(t *testing.T) {
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader_test.go]: nit: typo: s/convertation/conversion

File: plugin/storage/es/spanstore/reader_test.go [hunk 11]
```diff
@@ -702,12 +702,25 @@ func TestFindTraceIDs(t *testing.T) {
...
+	assert.NoError(t, err)
+	assert.Equal(t, 3, len(traceIDs))
+	assert.Equal(t, "1", traceIDs[0].String())
+
+	traceIDs, err = convertTraceIDsStringsToModels([]string{"dsfjsdklfjdsofdfsdbfkgbgoaemlrksdfbsdofgerjl"})
+	assert.EqualError(t, err, "Making traceID from string 'dsfjsdklfjdsofdfsdbfkgbgoaemlrksdfbsdofgerjl' failed: TraceID cannot be longer than 32 hex characters: dsfjsdklfjdsofdfsdbfkgbgoaemlrksdfbsdofgerjl")
+	assert.Equal(t, 0, len(traceIDs))
+}
+
+func TestTraceIDsModelsToStringsConvertation(t *testing.T) {
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader_test.go]: nit: typo: s/convertation/conversion

File: plugin/storage/es/spanstore/reader.go [hunk 12]
```diff
@@ -251,28 +251,37 @@ func (s *SpanReader) FindTraces(ctx context.Context, traceQuery *spanstore.Trace
...
-	return nil, errors.New("not implemented") // TODO: Implement
+	return convertTraceIDsStringsToModels(esTraceIDs)
 }
 
-func (s *SpanReader) multiRead(ctx context.Context, traceIDs []string, startTime, endTime time.Time) ([]*model.Trace, error) {
+func (s *SpanReader) multiRead(ctx context.Context, traceIDs []model.TraceID, startTime, endTime time.Time) ([]*model.Trace, error) {
 
 	childSpan, _ := opentracing.StartSpanFromContext(ctx, "multiRead")
-	childSpan.LogFields(otlog.Object("trace_ids", traceIDs))
+	childSpan.LogFields(otlog.Object("trace_ids", convertTraceIDsModelsToStrings(traceIDs)))
```
Comment (vprithvi) [inline plugin/storage/es/spanstore/reader.go]: Is this conversion necessary? The logging code is implemented as such:
```
func (ml fieldsAsMap) EmitObject(key string, value interface{}) {
	ml[key] = fmt.Sprintf("%+v", value)
}
```
which should print out traceIDs properly without doing additional work.

**Comment by vprithvi** [comment]
@vlamug It looks like the failure wasn't related to your change! Once the question about the string conversion is answered, we can merge!

**Comment by vlamug** [comment]
@vprithvi , thank you for review. The all needed changes were made.
